### PR TITLE
StateAPI for Verified Data Cap

### PR DESCRIPTION
**Pull request by dirkmc**
Fixes https://github.com/filecoin-project/lotus/issues/2222

### CODE REVIEW

File: cli/client.go [hunk 1]
```diff
@@ -316,6 +316,30 @@ var clientDealCmd = &cli.Command{
 			ref.TransferType = storagemarket.TTManual
 		}
 
+		// Check if the address is a verified client
+		dcap, err := api.StateVerifiedClientStatus(ctx, a, types.EmptyTSK)
+		if err != nil {
```
Comment (hannahhoward) [inline cli/client.go:321]: I believe err will be non-nil (because datacap key will be missing) when client is not verified.

File: cli/client.go [hunk 2]
```diff
@@ -316,6 +316,28 @@ var clientDealCmd = &cli.Command{
 			ref.TransferType = storagemarket.TTManual
 		}
 
+		// Check if the address is a verified client
+		_, err = api.StateVerifiedClientStatus(ctx, a, types.EmptyTSK)
+		isVerified := true
+		if err == lapi.NotFoundErr {
```
Comment (magik6k) [inline cli/client.go:322]: That won't work over JsonRPC.

File: node/impl/full/state.go [hunk 3]
```diff
@@ -810,3 +811,35 @@ func (a *StateAPI) StateMinerAvailableBalance(ctx context.Context, maddr address
...
+
+	vh, err := hamt.LoadNode(ctx, cst, st.VerifiedClients)
+	if err != nil {
+		return verifreg.DataCap{}, err
+	}
+
+	var dcap verifreg.DataCap
+	if err := vh.Find(ctx, string(addr.Bytes()), &dcap); err != nil {
+		if err == hamt.ErrNotFound {
+			return verifreg.DataCap{}, api.NotFoundErr
```
Comment (magik6k) [inline node/impl/full/state.go:839]: You can't compare errors when the method is invoked over JsonRPC. Instead this method could be returning the data cap as a pointer, with nil/null indicating 'not found'

**Comment by magik6k** [comment]
(pushed some formatting fixes)

File: api/api_full.go [hunk 4]
```diff
@@ -532,3 +535,5 @@ type Fault struct {
 	Miner address.Address
 	Epoch abi.ChainEpoch
 }
+
+var NotFoundErr = xerrors.Errorf("Not found")
```
Comment (magik6k) [inline api/api_full.go:539]: Unused now
```suggestion
```

File: api/api_full.go [hunk 5]
```diff
@@ -239,6 +241,7 @@ type FullNode interface {
 	StateGetReceipt(context.Context, cid.Cid, types.TipSetKey) (*types.MessageReceipt, error)
 	StateMinerSectorCount(context.Context, address.Address, types.TipSetKey) (MinerSectors, error)
 	StateCompute(context.Context, abi.ChainEpoch, []*types.Message, types.TipSetKey) (*ComputeStateOutput, error)
+	StateVerifiedClientStatus(ctx context.Context, addr address.Address, tsk types.TipSetKey) (*verifreg.DataCap, error)
```
Comment (magik6k) [inline api/api_full.go:244]: (copied comment from node/impl)

We auto-generate api docs from docstrings in `api/api*.go` files

```suggestion
	// StateVerifiedClientStatus returns the data cap for the given address.
	// Returns nil if there is no entry in the data cap table for the
	// address.
	StateVerifiedClientStatus(ctx context.Context, addr address.Address, tsk types.TipSetKey) (*verifreg.DataCap, error)
```

**Pull request merged**